man/man1/calendar.1
.TH CALENDAR 1
.SH NAME
calendar \- print upcoming events
.SH SYNOPSIS
.B calendar
[
.B \-y
]
[
.B \-p days
]
[
.I file ...
]
.SH DESCRIPTION
.I Calendar
reads the named files, default
.BR $HOME/lib/calendar ,
and writes to standard output any lines
containing today's or tomorrow's date.
Examples of recognized date formats are 
"4/11",
"April 11",
"Apr 11",
"11 April",
and
"11 Apr".
All comparisons are case insensitive.
.PP
If the
.B \-y
flag is given, an attempt is made to match on year too.  In this case,
dates of the forms listed above will be accepted if they are followed
by the current year (or last two digits thereof) or not a year —
digits not followed by white space or non-digits.
.PP
If the
.B \-p
flag is given, its argument is the number of days ahead to match
dates.  This flag is not repeatable, and it performs no special
processing at the end of the week.
.PP
On Friday and Saturday, events through Monday are printed.
.PP
To have your calendar mailed to you every day, use
.MR cron 8 .
.SH FILES
.TF $HOME/lib/calendar
.TP
.B $HOME/lib/calendar
personal calendar
.SH SOURCE
.B \*9/src/cmd/calendar.c


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/cmd/calendar.c
#include <u.h>
#include <libc.h>
#include <bio.h>
#include <regexp.h>
#include <ctype.h>

typedef struct Date	Date;
struct Date {
	Reprog *p;	/* an RE to match this date */
	Date *next;	/* pointer to next in list */
};

enum{
	Secondsperday = 24*60*60
};

Biobuf in;
int debug, matchyear;

Date *dates(Date**, Tm*);
void upper2lower(char*, char*, int);
void *alloc(unsigned int);

void
main(int argc, char *argv[])
{
	int i, fd, ahead;
	long now;
	char *line;
	Tm *tm;
	Date *first, *last, *d;
	char buf[1024];

	ahead = 0;
	ARGBEGIN{
	case 'y':
		matchyear = 1;
		break;
	case 'd':
		debug = 1;
		break;
	case 'p':
		ahead = atoi(ARGF());
		break;
	default:
		fprint(2, "usage: calendar [-y] [-d] [files ...]\n");
		exits("usage");
	}ARGEND;

	/* make a list of dates */
	now = time(0);
	tm = localtime(now);
	last = nil;
	first = dates(&last, tm);
	now += Secondsperday;
	tm = localtime(now);
	dates(&last, tm);
	if(tm->wday == 6){
		now += Secondsperday;
		tm = localtime(now);
		dates(&last, tm);
	}
	if(tm->wday == 0){
		now += Secondsperday;
		tm = localtime(now);
		dates(&last, tm);
	}
	if(ahead){
		now = time(0);
		now += ahead * Secondsperday;
		tm = localtime(now);
		dates(&last, tm);
	}

	for(i=0; i<argc || (i==0 && argc==0); i++){
		if(i==0 && argc==0)
			snprint(buf, sizeof(buf),
				"%s/lib/calendar", getenv("HOME"));
		else
			strcpy(buf, argv[i]);
		fd = open(buf, OREAD);
		if(fd<0 || Binit(&in, fd, OREAD)<0){
			fprint(2, "calendar: can't open %s: %r\n", buf);
			exits("open");
		}

		/* go through the file */
		while(line = Brdline(&in, '\n')){
			line[Blinelen(&in) - 1] = 0;
			upper2lower(buf, line, sizeof buf);
			for(d=first; d; d=d->next)
				if(regexec(d->p, buf, 0, 0)){
					print("%s\n", line);
					break;
				}
		}
		close(fd);
	}
	exits("");
}

char *months[] =
{
	"january",
	"february",
	"march",
	"april",
	"may",
	"june",
	"july",
	"august",
	"september",
	"october",
	"november",
	"december"
};

/*
 * Generate two Date structures.  First has month followed by day;
 * second has day followed by month.  Link them into list after
 * last, and return the first.
 */
Date*
dates(Date **last, Tm *tm)
{
	Date *first;
	Date *nd;
	char mo[128], buf[128];

	if(utflen(months[tm->mon]) > 3)
		snprint(mo, sizeof mo, "%3.3s(%s)?",
			months[tm->mon], months[tm->mon]+3);
	else
		snprint(mo, sizeof mo, "%3.3s", months[tm->mon]);
	if (matchyear)
		snprint(buf, sizeof buf,
			"(^| |\t)((%s( |\t)+)|(%d/))%d( |\t|$)(((%d|%d)( |\t|$))|[^0-9]|([0-9]+[^0-9 \t]))",
			mo, tm->mon+1, tm->mday, tm->year+1900, tm->year%100);
	else
		snprint(buf, sizeof buf,
			"(^| |\t)((%s( |\t)+)|(%d/))%d( |\t|$)",
			mo, tm->mon+1, tm->mday);
	if(debug)
		print("%s\n", buf);

	first = alloc(sizeof(Date));
	if(*last)
		(*last)->next = first;
	first->p = regcomp(buf);

	if (matchyear)
		snprint(buf, sizeof buf,
			"(^| |\t)%d( |\t)+(%s)( |\t|$)(((%d|%d)( |\t|$))|[^0-9]|([0-9]+[^0-9 \t]))",
			tm->mday, mo, tm->year+1900, tm->year%100);
	else
		snprint(buf, sizeof buf,
			"(^| |\t)%d( |\t)+(%s)( |\t|$)",
			tm->mday, mo);
	if(debug)
		print("%s\n", buf);
	nd = alloc(sizeof(Date));
	nd->p = regcomp(buf);
	nd->next = 0;
	first->next = nd;
	*last = nd;

	return first;
}

/*
 * Copy 'from' to 'to', converting to lower case
 */
void
upper2lower(char *to, char *from, int len)
{
	while(--len>0 && *from!='\0')
		*to++ = tolower((uchar)*from++);
	*to = 0;
}

/*
 * Call malloc and check for errors
 */
void*
alloc(unsigned int n)
{
	void *p;

	p = malloc(n);
	if(p == 0){
		fprint(2, "calendar: malloc failed: %r\n");
		exits("malloc");
	}
	return p;
}
